### ISSUE TITLE
Delegate whitelist filtering to Xdebug

### ISSUE THREAD

**Issue by sebastianbergmann** [sebastianbergmann/php-code-coverage#514]
Depends on https://bugs.xdebug.org/view.php?id=1059

**Comment by sebastianbergmann** [comment]
@bwoebi Will you add similar functionality to PHPDBG?

**Comment by sebastianbergmann** [comment]
https://bugs.xdebug.org/view.php?id=1059 has been implemented for Xdebug 2.6.0 and is available since Xdebug 2.6.0beta1.

As of 4723e24ad50181cd2e7aab67d9db92c3ff5db04b this should work, however it does not. Maybe @derickr can have a look? I may have made a mistake somewhere.

#### `master`
![screenshot from 2017-12-30 06-39-40](https://user-images.githubusercontent.com/25218/34451631-7a8f8a72-ed2c-11e7-8d5b-226465b5f8fe.png)
#### `issue-514`
![screenshot from 2017-12-30 06-54-57](https://user-images.githubusercontent.com/25218/34451693-7744f90e-ed2e-11e7-9d4a-ff765b5de8a5.png)

**Comment by derickr** [comment]
I think you're doing some weird things here:

- You seem to be setting the filter again for test run
- You give a long list of file names to the filter (40 items).
- You make PHP Code Coverage do work for seemingly all files, instead of just the ones from the white list now.
- I had a bug: 

The filter works best if you provide a few *directories* as filter and at the start. In order for it to work, it needs to "mark" oparrays as filtered/non-filtered, and the filter needs to be loaded *before* files/functions/classes are included in order to this at any speedy way.

In order to speed this up, making an auto-prepend file with:

```
<?php
xdebug_set_filter(XDEBUG_FILTER_CODE_COVERAGE, XDEBUG_PATH_WHITELIST, [ "/home/derick/dev/php/php-code-coverage/src" ]);
```

and running: ```php -dauto_prepend_file=/tmp/prepend.php vendor/bin/phpunit --coverage-html=/tmp/after-514```
```

**Comment by sebastianbergmann** [comment]
* Yes, the filter is currently set for each run. Further refactoring is required to avoid this, but if this has adverse effects beyond a performance penalty then that would be a bug in Xdebug, IMO.
* PHPUnit, and thus php-code-coverage, offer very flexible (one could argue too flexible) configuration of the whitelist. The result of reading and processing that configuration is a, possibly long, list of whitelisted files.

**Comment by derickr** [comment]
> - Yes, the filter is currently set for each run. Further refactoring is required to avoid this, but if this has adverse effects beyond a performance penalty then that would be a bug in Xdebug, IMO.

It doesn't have any adverse effects. *However*, the filter needs to be set-up before code is being included for the filter to act. There is no way to avoid this unless you are okay with having the filter check at every counting of code coverage line. That would be so much slower than we currently have, so not a good thing to have. The filter needs to be set-up *before* any user land code is included.

On top of that, *if* code is included before the filter is set-up, and only then the filter is set, it wouldn't do anything to filter out the already loaded code at all. This is why you suddenly get counted lines in for example the vendor/ and tests/ folders. To avoid that, while you don't set-up the filter before anything else runs, you *also* are going to use the white list filter you already previously had in PHP Code Coverage.

> - PHPUnit, and thus php-code-coverage, offer very flexible (one could argue too flexible) configuration of the whitelist. The result of reading and processing that configuration is a, possibly long, list of whitelisted files.

That is not a problem, it's just going to be a little slower when marking an oparray as being filtered. But that check is done only once per oparray, it shouldn't be a problem. However, if you can *avoid* this, and just set-up a filter based on path, then you gain even more performance.

**Comment by sebastianbergmann** [comment]
Closed in favour of #636.
